src/Client.cpp

```diff
@@ -1,245 +1,487 @@
-    request_received.data.reserve(256);
+int Client::getVersion()
+{
+    return msp_ver_;
 }
+    
 
-Client::~Client() {
-    for(const std::pair<msp::ID, msp::Request*> d : subscribed_requests)
-        delete d.second;
+void Client::setVariant(FirmwareVariant v) 
+{
+    fw_variant = v;
+}
 
-    for(const std::pair<msp::ID, SubscriptionBase*> s : subscriptions)
-        delete s.second;
+FirmwareVariant Client::getVariant() 
+{
+    return fw_variant;
 }
 
-void Client::connect(const std::string &device, const size_t baudrate) {
-    pimpl->port.open(device);
 
-    pimpl->port.set_option(asio::serial_port::baud_rate(baudrate));
-    pimpl->port.set_option(asio::serial_port::parity(asio::serial_port::parity::none));
-    pimpl->port.set_option(asio::serial_port::character_size(asio::serial_port::character_size(8)));
-    pimpl->port.set_option(asio::serial_port::stop_bits(asio::serial_port::stop_bits::one));
+bool Client::connect() 
+{
+    port.open(device_name_);
+    port.set_option(asio::serial_port::baud_rate(baud_rate_));
+    port.set_option(asio::serial_port::parity(asio::serial_port::parity::none));
+    port.set_option(asio::serial_port::character_size(asio::serial_port::character_size(8)));
+    port.set_option(asio::serial_port::stop_bits(asio::serial_port::stop_bits::one));
+    return start();
+    
 }
 
-void Client::start() {
-    thread = std::thread([this]{
-        running = true;
-        while(running) { processOneMessage(); }
-    });
+bool Client::disconnect()
+{
+    port.close();
+    return stop();
 }
 
-void Client::stop() {
-    running = false;
-    pimpl->io.stop();
-    pimpl->port.close();
-    thread.join();
-}
 
-bool Client::sendData(const uint8_t id, const ByteVector &data) {
-    std::lock_guard<std::mutex> lock(mutex_send);
+bool Client::start() 
+{
+    if (!port.is_open()) return false;
+    if (running_.test_and_set()) return false;
+    thread = std::thread([this]{
+        asio::async_read_until(port, buffer, 
+            std::bind(&Client::messageReady,this,std::placeholders::_1,std::placeholders::_2),
+            std::bind(&Client::processOneMessage,this,std::placeholders::_1,std::placeholders::_2));
+        io.run();
+        
+    });
+    return true;
+}
 
-    try {
-        asio::write(pimpl->port, asio::buffer("$M<",3));              // header
-        asio::write(pimpl->port, asio::buffer({uint8_t(data.size())})); // data size
-        asio::write(pimpl->port, asio::buffer({uint8_t(id)}));          // message id
-        asio::write(pimpl->port, asio::buffer(data));                   // data
-        asio::write(pimpl->port, asio::buffer({crc(id, data)}));        // crc
-    } catch (const asio::system_error &ec) {
-        if (ec.code() == asio::error::operation_aborted) {
-            //operation_aborted error probably means the client is being closed
-            return false;
+bool Client::stop() 
+{
+    bool rc = false;
+    if (running_.test_and_set()) {
```

Comment:
Why are you using a `std::atomic_flag` instead of `std::atomic<bool>`? An atomic bool would save you the clearing at the end and you could remove the `rc` variable and return directly.

Reply:
it makes things cleaner on the startup side by removing a race condition where two threads might both check the atomic and see it to be false, and then both try to set it true.